evm/p2p/discovery.py

```
@@ -0,0 +1,600 @@
+class DiscoveryProtocol(asyncio.DatagramProtocol):
+
+    """
+    ## Packet Data
+    All packets contain an `Expiration` date to guard against replay attacks.
+    The date should be interpreted as a UNIX timestamp.
+    The receiver should discard any packet whose `Expiration` value is in the past.
+    """
+    bootstrapped = False
+    transport = None
+    version = 4
+    expiration = 60  # let messages expire after N secondes
+    cmd_id_map = dict(ping=1, pong=2, find_node=3, neighbours=4)
+    rev_cmd_id_map = dict((v, k) for k, v in cmd_id_map.items())
+
+    # number of required top-level list elements for each cmd_id.
+    # elements beyond this length are trimmed.
+    cmd_elem_count_map = dict(ping=4, pong=3, find_node=2, neighbours=2)
+
+    encoders = dict(cmd_id=chr,
+                    expiration=rlp.sedes.big_endian_int.serialize)
+
+    decoders = dict(cmd_id=safe_ord,
+                    expiration=rlp.sedes.big_endian_int.deserialize)
+
+    def __init__(self, privkey, address, bootstrap_nodes):
+        self.privkey = privkey
+        self.pubkey = private_key_to_public_key(self.privkey)
+        self.address = address
+        self.bootstrap_nodes = bootstrap_nodes
+        self.this_node = Node(self.pubkey, address)
+        self.nodes = LRUCache(2048)   # nodeid->Node,  fixme should be loaded
+        self.kademlia = kademlia.KademliaProtocol(self.this_node, wire=self)
+        self.nat_upnp = add_portmap(address.udp_port, 'UDP', 'Ethereum DEVP2P Discovery')
+        log.info('starting discovery proto on', this=self.this_node)
+
+    def listen(self, loop):
+        return loop.create_datagram_endpoint(
+            lambda: self, local_addr=(self.address.ip, self.address.udp_port))
+
+    def connection_made(self, transport):
+        self.transport = transport
+
+    @asyncio.coroutine
+    def bootstrap(self):
+        if self.bootstrapped or len(self.bootstrap_nodes) == 0:
```

These both feel like anti-patterns.

* The `self.bootstrapped` state is something that I'd like to look for an alternate solution which didn't require storing state (low priority)
* The *silent* return  if the list of bootstrap nodes feels like a silent failure.

I don't have enough knowledge yet to know if this is a dumb solution but...

What do you think about modifying this so that it instead takes an iterable of bootstrap nodes as an argument.  This would remove the silent return when there are no bootstrap nodes and allow moving that check higher in the stack to allow the `DiscoveryProtocol` object to not care about that stuff as well as possibly removing the need to keep the `self.bootstrapped` state around?